casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -70,113 +56,209 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Log: Time: SafetyOracle: Las
+                           )
+                         }
+                   } yield r
+               }
+    } yield result
+  }
 
-    def add: F[ValidBlockProcessing] = spanF.trace(AddBlockMetricsSource) {
-      for {
-        _      <- spanF.mark("block-store-put")
-        dag    <- blockDag
-        status <- internalAddBlock(b, dag)
-        _      <- spanF.mark("block-added-status")
-      } yield status
-    }
+  private def safetyRangeFilled: F[Boolean] = {
+    import cats.instances.list._
+    for {
+      approvedBlock  <- getGenesis
+      abHasLowHeight = ProtoUtil.blockNumber(approvedBlock) < expirationThreshold
 
-    import cats.instances.either._
+      // validators as per approved block state
+      abValidators = approvedBlock.body.state.bonds.map(_.validator)
+
+      cbPendants       <- casperBuffer.getPendants
+      cbPendantsBlocks <- cbPendants.map(ProtoUtil.getBlock(_)).toList.sequence
+
+      enoughBlocksAreInBuffer = abValidators.foldLeft(true)((acc, validator) => {
+        val pendantFromValidator = cbPendantsBlocks.find(b => b.sender.equals(validator))
+        val enoughBlocksAreInBufferForValidator = pendantFromValidator.exists(
+          p =>
+            ProtoUtil.blockNumber(p) <= ProtoUtil.blockNumber(approvedBlock) - expirationThreshold
+        )
+        acc && enoughBlocksAreInBufferForValidator
+      })
+
+      bufferFilled = abHasLowHeight || enoughBlocksAreInBuffer
+    } yield bufferFilled
+  }
+
+  private def isApprovedBlockChild(blockMessage: BlockMessage): F[Boolean] =
     for {
-      // Save block in block store before waiting for blockProcessingLock
-      // This is called inside block message handler as well, but required here in case block is not received from peer
-      // but proposed by node itself.
-      _ <- BlockStore[F].contains(b.blockHash).ifM(().pure[F], BlockStore[F].put(b))
-      _ <- Log[F].info(
-            s"Block ${PrettyPrinter.buildString(b.blockHash)} asks for blockProcessingLock."
-          )
-      // ATM Casper is allowed to add only one block at a time
-      status <- blockProcessingLock.withPermit {
-                 val exists = for {
+      parents <- ProtoUtil.getParents(blockMessage)
+      r = parents.foldLeft(false)(
+        (acc, parent) => (acc || parent.blockHash.equals(approvedBlock.blockHash))
+      )
+    } yield r
+
+  /**
+    * Add block to Casper
+    *
+    * @param b                  Block
+    * @param allowAddFromBuffer If block is in CaspeBuffer, try to add it instead of reporting already processed.
+    *                           This is required for continuing syncing after node restart
+    */
+  def addBlock(b: BlockMessage, allowAddFromBuffer: Boolean): F[ValidBlockProcessing] =
+    for {
+      // We force Casper to be single threaded. More investigations needed to allow it be multi threaded.
+      casperLocked <- ~^(blockProcessingLock.tryAcquireN(1L))
```

Comment:
This is not safe if error is thrown and _release_ is never called.

Bracket can ensure that release is called always like with `try...finally`. We have similar situation in Block API.
https://github.com/rchain/rchain/blob/8fb7bc466ea9a7b516133da40b85da95d11595bb/casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala#L86-L87